op-e2e/system_fpp_test.go

```diff
@@ -27,6 +27,114 @@ func TestVerifyL2OutputRootDetached(t *testing.T) {
 	testVerifyL2OutputRoot(t, true)
 }
 
+func TestVerifyL2OutputRootEmptyBlock(t *testing.T) {
+	testVerifyL2OutputRootEmptyBlock(t, false)
+}
+
+func TestVerifyL2OutputRootEmptyBlockDetached(t *testing.T) {
+	testVerifyL2OutputRootEmptyBlock(t, true)
+}
+
+// TestVerifyL2OutputRoot_EmptyBlock asserts that the program can verify the output root of an empty block
+// induced by missing batches.
+// Setup is as follows:
+// - create initial conditions and agreed l2 state
+// - stop the batch submitter to induce empty blocks
+// - wait for the seq window to expire so we can observe empty blocks
+// - select an empty block as our claim
+// - reboot the batch submitter
+// - update the state root via a tx
+// - run program
+func testVerifyL2OutputRootEmptyBlock(t *testing.T, detached bool) {
+	InitParallel(t)
+	ctx := context.Background()
+
+	cfg := DefaultSystemConfig(t)
+	// Use a small sequencer window size to avoid test timeout while waiting for empty blocks
+	// But not too small to ensure that the updated state after claim selection is published
+	cfg.DeployConfig.SequencerWindowSize = 16
+
+	sys, err := cfg.Start()
+	require.Nil(t, err, "Error starting up system")
+	defer sys.Close()
+
+	log := testlog.Logger(t, log.LvlInfo)
+	log.Info("genesis", "l2", sys.RollupConfig.Genesis.L2, "l1", sys.RollupConfig.Genesis.L1, "l2_time", sys.RollupConfig.Genesis.L2Time)
+
+	l1Client := sys.Clients["l1"]
+	l2Seq := sys.Clients["sequencer"]
+	l2Verif := sys.Clients["verifier"]
+	rollupRPCClient, err := rpc.DialContext(context.Background(), sys.RollupNodes["sequencer"].HTTPEndpoint())
+	require.Nil(t, err)
+	rollupClient := sources.NewRollupClient(client.NewBaseRPCClient(rollupRPCClient))
+
+	verifierRPCClient, err := rpc.DialContext(context.Background(), sys.RollupNodes["sequencer"].HTTPEndpoint())
+	require.Nil(t, err)
+	verifierClient := sources.NewRollupClient(client.NewBaseRPCClient(verifierRPCClient))
+
+	t.Log("Sending transactions to setup existing state, prior to challenged period")
+	aliceKey := cfg.Secrets.Alice
+	receipt := SendL2Tx(t, cfg, l2Seq, aliceKey, func(opts *TxOpts) {
+		opts.ToAddr = &cfg.Secrets.Addresses().Bob
+		opts.Value = big.NewInt(1_000)
+	})
+	nonEmptyBlock, err := waitForBlock(receipt.BlockNumber, l2Verif, time.Duration(2*cfg.DeployConfig.L1BlockTime)*time.Second)
+	require.Nil(t, err, "Waiting for block on verifier")
+	require.Equal(t, receipt.BlockHash, nonEmptyBlock.Hash(), "send tx should be published")
+
+	t.Log("Capture current L2 head as agreed starting point")
+	l2AgreedBlock, err := l2Seq.BlockByNumber(ctx, nil)
+	require.NoError(t, err, "could not retrieve l2 agreed block")
+	l2Head := l2AgreedBlock.Hash()
+
+	t.Log("=====Stopping batch submitter=====")
+	err = sys.BatchSubmitter.Stop(ctx)
```

Comment:
Feels like we should wait for the safe head to update to our new block before stopping the batch submitter. Otherwise we might stop the batcher before it has published the tx we sent.